### PR TITLE
Add AppInsights heartbeat support

### PR DESCRIPTION
**Pull request by xavierdecoster**
Required for NuGet/Engineering#2537

### CODE REVIEW

File: src/NuGet.Services.Logging/ApplicationInsights.cs [hunk 1]
```diff
@@ -1,27 +1,86 @@
...
+
+        private static void InitializeTelemetryConfiguration(string instrumentationKey, TimeSpan? heartbeatInterval)
         {
             if (!string.IsNullOrWhiteSpace(instrumentationKey))
             {
                 TelemetryConfiguration.Active.InstrumentationKey = instrumentationKey;
                 TelemetryConfiguration.Active.TelemetryInitializers.Add(new TelemetryContextInitializer());
 
+                // Configure heartbeat interval if specified.
+                // When not defined, the DiagnosticsTelemetryModule will use a default value of 15 minutes.
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/ApplicationInsights.cs]: Hmm, good point, it is possible, and in fact I should set `IsHeartbeatEnabled=false` in that case. Will fix.

File: src/NuGet.Services.Logging/NuGet.Services.Logging.csproj [hunk 2]
```diff
@@ -66,7 +66,7 @@
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.ApplicationInsights">
-      <Version>2.3.0</Version>
+      <Version>2.10.0</Version>
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/NuGet.Services.Logging.csproj]: Heartbeat was only added in 2.5.x with some fixes related to heartbeat in 2.6... Sounds like we need to bite the bullet, scan for problems, and fix them.

File: src/NuGet.Services.Logging/ApplicationInsights.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/ApplicationInsights.cs]: @skofman1 Actually, if no heartbeat interval is configured, and the value is `null`, shouldn't we keep it enabled and use default value of 15 minutes? The hard-coded default in Application Insights is 15 minutes when no value is configured, and heartbeat is enabled by default.

File: src/NuGet.Services.Logging/ApplicationInsights.cs [hunk 3]
```diff
@@ -1,27 +1,86 @@
...
+                        if (module is IHeartbeatPropertyManager heartbeatManager)
+                        {
+                            HeartbeatManager = heartbeatManager;
+                        }
+                    }
+                }
+                catch (Exception hearbeatManagerAccessException)
+                {
+                    // An non-critical, unexpected exception occurred trying to access the heartbeat manager.
+                    Trace.TraceError(hearbeatManagerAccessException.ToInvariantString());
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/ApplicationInsights.cs]: Fixed in https://github.com/NuGet/ServerCommon/pull/308/commits/b87320a6c207649d99eb002f92c10f82fed3b140

File: src/NuGet.Services.Logging/ApplicationInsights.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/ApplicationInsights.cs]: Updated the existing comment a little :) https://github.com/NuGet/ServerCommon/pull/308/commits/b87320a6c207649d99eb002f92c10f82fed3b140

File: src/NuGet.Services.Logging/NuGet.Services.Logging.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/NuGet.Services.Logging.csproj]: @joelverhagen do you have memory or history of how you detected the perf/telemetry issues? Or did you open issues against Application Insights to report the regression you experienced?

Would like to learn how to proceed on this issue, as this PR is required to enable heartbeats, and a is blocking changes required in `NuGet.Jobs` to enable heartbeats: I need updated `NuGet.Services.*.nupkg` packages including the changes in this PR both for integrating the change as well as testing for potential telemetry/perf issues.

**Comment by xavierdecoster** [comment]
Created a [pre-release build and artifacts for this PR](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=2995133&view=results) to facilitate testing the changes in consuming projects.

**Comment by xavierdecoster** [comment]
Additional complexity: our custom telemetry enrichers are relying on an obsolete `telemetry.Context.Properties` API. We'll also need to change these as part of the AI upgrade in the gallery.
https://stackoverflow.com/questions/52821523/application-insights-how-to-use-isupportproperties-in-telemetryinitializer

File: src/NuGet.Services.Logging/NuGet.Services.Logging.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (cristinamanum) [reply src/NuGet.Services.Logging/NuGet.Services.Logging.csproj]: The move to Application Insights 2.10.0 in Gallery caused a Sev 2 incident. Postmortem 268899 has details. The doc attached to the postmortem has additional information regarding the RCA. The PR https://github.com/NuGet/NuGetGallery/pull/7292/files reverts the change that caused the incident.

[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGet.Services.Logging/NuGet.Services.Logging.csproj]: Yes, but I'm wondering why the changes to the `applicationinsights.config` file were made. Those are not autogenerated by the update of the ApplicationInsights SDK, so they've been done manually. Why were those changes introduced? See [this comment](https://github.com/NuGet/NuGetGallery/pull/7253/files/f50db471f11f88550e74c801f5f9b717ca2a0add#r320259162) on that PR.

**Comment by xavierdecoster** [comment]
Given the load tests on NuGetGallery with AppInsights v2.10 looked good, can I go ahead and merge this in? We'll also need a release of ServerCommon to be able to consume it elsewhere.

@skofman1 @joelverhagen

**Comment by skofman1** [comment]
Feel free to merge :shipit:

**Pull request merged**